### ISSUE TITLE
changing the length of build_files of new_* rules fails

### ISSUE THREAD

**Issue by bsilver8192** [bazelbuild/bazel#566]
After changing the length of the build file for a new_* rule, Bazel re-reads the old length of bytes from the file. If it got smaller, this fails. If the file got longer, the bytes beyond the original length are ignored.

I think this is related to @haxorz's recent changes to read/parse/etc BUILD files less often.

For example, starting with the following files:
WORKSPACE:
```python
new_local_repository(
  name = 'test_repo',
  path = '/tmp/test_repo',
  build_file = 'test.BUILD',
)
```
test.BUILD:
```python
cc_binary(
  name = 'test',
  srcs = ['test.cc'],
)
```
/tmp/test_repo/test.cc:
```c++
int main() { return 0; }
```

`bazel build @test_repo//:test` succeeds as expected. However, running `echo foobar >> test.BUILD` and then `bazel build @test_repo//:test` again succeeds even though it shouldn't. It fails like it should after a `bazel shutdown`.

**Comment by bsilver8192** [comment]
Correction: this has been around since at least 47182329400e359a7a6ad5857cf1854a4f9ec720, so it looks like those recent changes were responsible.

Also, I thought about it and realized there's an easy way to test this, so here's a patch for a test to commit along with the fix:
```patch
diff --git a/src/test/shell/bazel/external_integration_test.sh b/src/test/shell/bazel/external_integration_test.sh
index 6421d81..24a9eb8 100755
--- a/src/test/shell/bazel/external_integration_test.sh
+++ b/src/test/shell/bazel/external_integration_test.sh
@@ -639,4 +639,34 @@ EOF
   assert_contains "def" bazel-genfiles/external/x/catter.out
 }
 
+function test_editing_build_file() {
+  echo "abc" > w
+  tar czf x.tar.gz w
+  local sha256=$(sha256sum x.tar.gz | cut -f 1 -d ' ')
+  serve_file x.tar.gz
+
+  cat > WORKSPACE <<EOF
+new_http_archive(
+    name = "x",
+    url = "http://localhost:$nc_port/x.tar.gz",
+    sha256 = "$sha256",
+    build_file = "x.BUILD",
+)
+EOF
+  cat > x.BUILD <<EOF
+genrule(
+    name = "catter",
+    cmd = "cat \$< > \$@",
+    outs = ["catter.out"],
+    srcs = ["w"],
+)
+EOF
+
+  bazel build @x//:catter || fail "Build failed"
+  assert_contains "abc" bazel-genfiles/external/x/catter.out
+  echo "A syntax error!" >> x.BUILD
+  bazel build @x//:catter && fail "Build should have failed"
+  expect_log "syntax error"
+}
+
 run_suite "external tests"
```

**Comment by kchodorow** [comment]
Thanks for the repro case!  I'm on an external repos fixit binge, so if #352 doesn't fix it (changelist in progress), this will be the next thing I tackle.

**Comment by AustinSchuh** [comment]
Did this go anywhere?  I've been working on BUILD files for external repos recently, and my 'bazel shutdown; bazel build //my:target.' workflow is getting old...

**Comment by kchodorow** [comment]
I've been working on the commit that would not end (I started the code review on September 8th) that will unblock this.  I think I'm close to an LGTM now (although I've thought so in the past, too) but hopefully this week it'll be in and I can get this fixed before Christmas.

**Comment by kchodorow** [comment]
This was fixed in f9fdc8dfced8b2b14561720623126a91e04b22cb and I think https://github.com/bazelbuild/bazel/blob/master/src/test/shell/bazel/external_correctness_test.sh#L54-L68 covers the test case above.
